src/nupic/ntypes/ArrayBase.cpp

```diff
@@ -233,176 +254,232 @@ void ArrayBase::setCount(size_t count) {
     return false;
-  return (buffer_ == a.buffer_);
+  return (buffer_.get() == a.buffer_.get());
 }
 
-template <typename T> static void NonZeroT(ArrayBase& a, const ArrayBase* b) {
+template <typename T>
+static void NonZeroT(ArrayBase &a, const ArrayBase *b, size_t offset) {
+  // Helper function for toSparse().
   // populate the new array in a with indexes of non-zero values in b.
   // NOTE: NTA_BasicType_SDR and NTA_BasicType_Sparse not handled here.
-  a.allocateBuffer(b->getCount()); // allocating more space than we need.
   T *from = (T *)b->getBuffer();
-  UInt32 j = 0u;
+  size_t j = a.getCount(); // current end of buffer.
   UInt32 *Destptr = (UInt32 *)a.getBuffer();
-  for (UInt32 i = 0u; i < b->getCount(); i++) {
+  for (size_t i = 0u; i < b->getCount(); i++) {
     if (from[i])
-      Destptr[j++] = (UInt32)i;
+      Destptr[j++] = (UInt32)i + (UInt32)offset;
   }
-  a.setCount(j); // set the size.
+  a.setCount(j); // set new end of buffer.
 }
 
 // populate the given array a with the a sparse version of the current array.
+// Destination buffer already allocated and checked for size.  Could be called
+// multiple times to populate different offsets in the buffer. Each new index
+// is appended to the buffer.
 void ArrayBase::toSparse(ArrayBase &a, size_t offset) const {
-  UInt32 *p1, *p2;
+  UInt32 *p1;
+  const UInt32 *p2;
 
-  a.type_ = NTA_BasicType_Sparse;
-  switch (type_) {
+  a.type_ = NTA_BasicType_Sparse; // to Array of type Sparse
+  if (!has_buffer()) {
+    a.setCount(0); // source buffer is empty.
+    return;
+  }
+  switch (type_) { // coming from Array of this type
   case NTA_BasicType_Byte:
-    NonZeroT<Byte>(a, this);
+    NonZeroT<Byte>(a, this, offset);
     break;
   case NTA_BasicType_Int16:
-    NonZeroT<Int16>(a, this);
+    NonZeroT<Int16>(a, this, offset);
     break;
   case NTA_BasicType_UInt16:
-    NonZeroT<UInt16>(a, this);
+    NonZeroT<UInt16>(a, this, offset);
     break;
   case NTA_BasicType_Int32:
-    NonZeroT<Int32>(a, this);
+    NonZeroT<Int32>(a, this, offset);
     break;
   case NTA_BasicType_UInt32:
-    NonZeroT<UInt32>(a, this);
+    NonZeroT<UInt32>(a, this, offset);
+    break;
+  case NTA_BasicType_Int64:
+    NonZeroT<Int64>(a, this, offset);
+    break;
+  case NTA_BasicType_UInt64:
+    NonZeroT<UInt64>(a, this, offset);
     break;
   case NTA_BasicType_Real32:
-    NonZeroT<Real32>(a, this);
+    NonZeroT<Real32>(a, this, offset);
     break;
   case NTA_BasicType_Real64:
-    NonZeroT<Real64>(a, this);
+    NonZeroT<Real64>(a, this, offset);
     break;
   case NTA_BasicType_Bool:
-    NonZeroT<bool>(a, this);
-    break;
-  case NTA_BasicType_SDR: 
-    {
-      SDR_flatSparse_t &v = getSDR()->getFlatSparse();
-      a.allocateBuffer(v.size());
-      p1 = (UInt32 *)a.getBuffer();
-      p2 = v.data();
-      for (size_t i = 0u; i < v.size(); i++) {
-        *p1++ = *p2++;
-      }
-    //getSDR()->setFlatSparseInplace();
-    }
+    NonZeroT<bool>(a, this, offset);
```

Comment:
instead of all these switch-case, can you do `NonZeroT<type_>(a, this, offset)` ?